File: morse/js/board.js
/*
    This file is part of Morse.

    Morse is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Morse is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Morse.  If not, see <http://www.gnu.org/licenses/>.
*/

$(document).on("ready", function () {

  function discoverVisibleTopics (){

    /* 
       discoverVisibleTopics() looks for topic items in the current 
       viewport and marks them as discovered via ajax. this include 
       removing the css fresh class
       */

    var topicItems = $(".topic-item").filter(function(){ 
      return $(this).visible(); 
    });

    /* collect ids of visible posts */
    var topicIDs = Array();
    topicItems.each(function () {
      topicID = $(this).attr("topic-id");
      topicIDs.push(topicID);
    });

    var data = new Object({topicIDs: topicIDs});
    var json = $.toJSON(data);

    /* only send, if topics were found */
    if (topicIDs.length > 0){
      $.ajax({
        url: "/topics/discover",
        data: json,
        error: handleAjaxErrorBy( alertGlobal ),
        success: function () { topicItems.removeClass("fresh"); },
      });
    }

  }
  function rebindTopicItemEvents () {

  $(".close-topic").off("click");
  $(".close-topic").on("click", function(){
    var topicID = $(this).parents(".topicitem").attr("topic-id");

    $.ajax({
      url: "/topic/" + topicID + "/close",
      error: handleAjaxErrorBy( alertGlobal ),
      success: function(response){
        topicID = response.closedID;
        var item = $(".topicitem[topic-id=\"" + topicID + "\"]");
        item.addClass("topicclosed");
        item.find(".close-topic").fadeOut(0);
        item.find(".reopen-topic").fadeIn(200);
      },
    });
  });

  $(".reopen-topic").off("click");
  $(".reopen-topic").on("click", function(){
    var topicID = $(this).parents(".topicitem").attr("topic-id");

    $.ajax({
      url: "/topic/" + topicID + "/reopen",
      error: handleAjaxErrorBy( alertGlobal ),
      success: function(response){
        topicID = response.openedID;
        var item = $(".topicitem[topic-id=\"" + topicID + "\"]");
        item.removeClass("topicclosed");
        item.find(".reopen-topic").fadeOut(0);
        item.find(".close-topic").fadeIn(200);
      },
    });
  });

  }

  /* follow switch */
  $("#follow-board").click(function(){
    var boardID = $("#board").attr("board-id");
    var that = $(this);

    $.ajax({
      url: "/board/" + boardID + "/follow",
      error: handleAjaxErrorBy( alertGlobal ),
      success: function(response){
        switchButton(that);
      },
    });
  });

  $("#unfollow-board").click(function(){
    var boardID = $("#board").attr("board-id");
    var that = $(this);

    $.ajax({
      url: "/board/" + boardID + "/unfollow",
      error: handleAjaxErrorBy( alertGlobal ),
      success: function(response){
        switchButton(that);
      },
    });
  });

  $(".order-option").click(function(){
    if ($(this).hasClass("option-selected")){
      var switched = $(this).attr("switched");
      var html = $(this).html();
      console.log(switched);
      $(this).attr("switched", html );
      $(this).html( switched );
    }
    $(".order-option").removeClass("option-selected");
    $(this).addClass("option-selected");
    $("#active-order").html( $(this).html() );
  });

  $("#new-topic-button").click(function(){
    showNewContributionField();
  });

  $("#top-action").mouseenter(function(){

    /* block if divs are still in motion */
    var currentTop = parseInt($("#board-title").css("top"));
    if (currentTop > 0) { return true; }

    var height = $("#top-action a").height();
    var timer = setTimeout(function(){
      $("#board").animate( { top : height + "px" }, {duration: 500 });
      $("#board-description").animate( { top : height + "px" }, {duration: 500 });
      $("#board-title").animate( { top : height + "px" }, {duration: 500, complete: fadeInTopActionInfo });
    }, 400);
    $(this).data("toolRevealDelay", timer);

  });  

  function fadeInTopActionInfo (){
    $("#top-action a").fadeIn(200);
  }

  $("#top-action").mouseleave(function(){

    /* clear timer, that was triggered on mousenter */
    var timer = $(this).data("toolRevealDelay");
    clearTimeout(timer);

    fadeOutTopActionInfo();
    $("#board-title").animate( { top : "0" }, {duration: 500});
    $("#board-description").animate( { top : "0" }, {duration: 500});
    $("#board").animate( { top : "0" }, {duration: 500});

  });  

  function fadeOutTopActionInfo (){
    $("#top-action a").fadeOut(200);
  }

  $("#board-title").swipeSnap( { target: "top", snapInfo: $("#snap-tooltip-top"), 
                                 companions: [$("#board"), $("#board-description")], 
                                 callback: gotoBoardIndex });

  function gotoBoardIndex (){
    window.location = "/";
  }

  /* bidirectional infinite scrolling */
  var slot = "/board/" + $("#board").attr("board-id") + "/topics.json";
  var builder = "/board/" + $("#board").attr("board-id") + "/certaintopics";
  var scrolling = new InfiniteScrolling();
  scrolling.init(".topic-item", slot, builder, [rebindToolTipEvents, rebindTopicItemEvents, 
                                               checkForUnreadPosts, discoverVisibleTopics], 3000, 20, 5);
  scrolling.showFromStart();
 
  $(window).scroll(scrolling.scroll, discoverVisibleTopics);

  $("#itemcontainer").on("emptycontainer", function(){
    $("#no-topics-yet").fadeIn(200);
  });

  $(".filter-option .checkbox").change(function(){
    var filterStatus = [];
    $(".filter-option .checkbox").each(function(){
      var id = $(this).attr("id");
      var active = $(this).is(":checked");
      filterStatus.push( [id, active] );
    });
    var data = new Object({filterStatus: filterStatus});
    var json = $.toJSON(data);
    $.ajax({
      url: "/filter/topics",
      data: json,
      error: handleAjaxErrorBy( alertGlobal ),
      success: function(){
        $("#no-topics-yet").fadeOut(0);
	    scrolling.showFromStart();
        updateTopicCache();
      },
    });
  });

  function updateTopicCache (){
    var boardID = $("#board").attr("board-id");
    $.ajax({
      url: "/board/" + boardID + "/update-topic-cache",
      data: json,
      error: handleAjaxErrorBy( alertGlobal ),
    });
  }

  function checkForUnreadPosts (){ 
    $(".topic-item").each(function(){
      var topicItem = $(this);
      if (topicItem.attr("topic-followed") === "True"){
        var topicID = topicItem.attr("topic-id");
        $.ajax({
          url: "/topic/" + topicID + "/unread.json",
          error: handleAjaxErrorBy( alertGlobal ),
          type: "GET", 
          success: function(response){
            updateSideCounter(topicItem, response.unreadCount); 
          },
        });
      }
    });   
 
    setTimeout(checkForUnreadPosts, 5000);
  }

  function updateSideCounter (topicItem, count){
    var counter = topicItem.children(".item-side-counter");
    if (parseInt(counter.html()) === count){
      return true;
    }
    counter.html(count);
    counter.fadeIn(500);
  }

});

function showNewContributionField (){
  $("#new-topic-button").slideUp(400);
  $("#new-contribution").slideDown(400);
  $("#new-topic-title").focus();
}

File: morse/slots/board.py
#!/usr/bin/python

#    This file is part of Morse.
#
#    Morse is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    Morse is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with Morse.  If not, see <http://www.gnu.org/licenses/>.

from . import app
from flask.ext.login import current_user, login_required
from flask import jsonify, request
from ..rights import check_ban, possibly_banned
from ..validators import json_input, String
from ..protocols import ajax_triggered
from ..models import db
from ..models.core import Board
from ..models.discussion import Topic, Post, DiscoveredTopic
from ..generators import TopicListGenerator
from ..wrappers import TopicWrapper, PostWrapper, BoardWrapper
from ..cache import TopicCache
from sqlalchemy import not_

@app.route('/board/<board_str>/refresh-topic-cache', methods=['GET'])
@ajax_triggered
def refresh_topic_cache (board_str):
    """ 
    updates the topic cache for current user. this function
    is called via ajax in board.js when filter or sorting
    options change
    """
    board_id = int(board_str.split("-")[0])

    board = Board.query.get(board_id)
    if not board:
        return "boardnotfound", 404

    cache = TopicCache()
    cache.refresh(board_id)
    return ""

@app.route('/board/<board_str>/topics.json', methods=['GET'])
@ajax_triggered
def get_topics (board_str):
    """ 
    returns a JSON object with structure {topicIDs: [topic_id]}
    this function is called via ajax in board.js
    """
    board_id = int(board_str.split("-")[0])

    board = Board.query.get(board_id)
    if not board:
        return "boardnotfound", 404

    if not current_user.may_read(board):
        return "forbidden", 403

    topic_ids = []
    for topic in TopicListGenerator(board_id):
        topic_ids.append(topic.id)

    cache = TopicCache()
    cache.refresh(board_id) 

    return jsonify(IDs = topic_ids)

@app.route('/board/<board_str>/start-topic', methods=['POST'])
@possibly_banned
@ajax_triggered
@json_input({"title": String(), "text": String()})
def start_topic (board_str):
    """ 
    Creates a new topic. Is called by the javascript event
    callback for #docreate
    :param board_id: board in which the new topic 
                     should be started
    :rtype: json 
    """
    board_id = int(board_str.split("-")[0])
    check_ban(board_id)

    board = Board.query.get(board_id)
    if not board:
        return "boardnotfound", 404

    if not current_user.may_post_in(board):
        return "forbidden", 403 

    title = request.json["title"]
    text = request.json["text"]

    topic = Topic(board_id, title)
    db.session.add(topic)
    db.session.commit()

    if not current_user.is_anonymous():
        wrapper = TopicWrapper(topic)
        wrapper.follow()

    post = Post(current_user.id, text, topic.id, request.remote_addr)
    db.session.add(post)
    db.session.commit()
  
    post.calculate_traits() 
    db.session.commit()

    post.observe_traits()
    db.session.commit()

    topic.calculate_interesting_value()
    db.session.commit()

    if not current_user.is_anonymous():
        post = PostWrapper(post)
        post.read()

    cache = TopicCache()
    cache.refresh(board_id)

    return jsonify(topicId = topic.id)

@app.route('/board/<board_str>/follow', methods=['POST'])
@login_required
@ajax_triggered
def follow_board (board_str):
    """ 
    follows a board. this function is called via ajax
    by the click event handler for #follow-board in board.js
    """
    board_id = int(board_str.split("-")[0])

    board = Board.query.get(board_id)
    if not board:
        return "nosuchboard", 400
 
    board = BoardWrapper(board)
    board.follow()
    return ""

@app.route('/board/<board_str>/unfollow', methods=['POST'])
@login_required
@ajax_triggered
def unfollow_board (board_str):
    """ 
    unfollows a board. this function is called via ajax
    by the click event handler for #follow-board in board.js
    """
    board_id = int(board_str.split("-")[0])

    board = Board.query.get(board_id)
    if not board:
        return "nosuchboard", 400

    board = BoardWrapper(board)
    board.unfollow()
    return ""

@app.route('/board/<board_str>/undiscovered.json', methods=['GET'])
@ajax_triggered
def undiscovered_topic_stats (board_str):
    """ 
    returns a JSON undiscovered topic stats object for a board defined by board_str
    (structure {unreadCount, firstUnreadID}) this function is called via ajax in index.js
    """
    board_id = int(board_str.split("-")[0])

    board = Board.query.get(board_id)
    if not board:
        return "boardnotfound", 404

    if current_user.is_anonymous():
        return jsonify(undiscoveredCount = 0, firstUndiscoveredID = None)    

    topic_id_generator = Topic.query.filter(Topic.board_id == board_id).values(Topic.id)
    topic_ids = [oneple[0] for oneple in topic_id_generator] 

    discovered_ids_generator = DiscoveredTopic.query.filter(DiscoveredTopic.user_id == current_user.id, 
                                                            DiscoveredTopic.topic_id.in_(topic_ids)).values(DiscoveredTopic.topic_id)
    discovered_ids = [oneple[0] for oneple in discovered_ids_generator] 
    
    undiscovered_count = Topic.query.filter(Topic.board_id == board_id, not_(Topic.id.in_(discovered_ids))).count()
    first_undiscovered = Topic.query.filter(Topic.board_id == board_id, not_(Topic.id.in_(discovered_ids))).first()

    if not first_undiscovered:
        first_undiscovered_id = None
    else:
        first_undiscovered_id = first_undiscovered.id

    return jsonify(undiscoveredCount = undiscovered_count, firstUndiscoveredID = first_undiscovered_id)    
